### ISSUE TITLE
Preferred means of consumption?

### ISSUE THREAD

**Issue by tintoy** [OmniSharp/csharp-language-server-protocol#17]
Hi.

I built an [extension](https://github.com/tintoy/msbuild-project-tools-vscode) for VS Code to provide completions for `PackageReference` elements but, going forward, I'd like to extend it by [building an MSBuild language server](https://github.com/tintoy/msbuild-project-tools-vscode/issues/2). This project looks to be exactly what I'm looking for :)

I see that the README states this is still a work-in-progress so I'm not expecting a stable release channel right now but I'd appreciate it if you could tell me you intend for people to consume this framework - is there a public or private package feed I could consume?

Thanks,

Adam.

**Comment by david-driscoll** [comment]
I have not yet published to nuget yet... but I'll look at doing that tonight if I can finish what I'm working on at the moment.

The package is however available via the appveyor nuget feed https://ci.appveyor.com/nuget/csharp-language-server-protocol.

In terms of being a work in progress.  @Lercher has had success with it afaik, and I'll be taking a deep dive into consumption in the next week or two, as I implement LSP into omnisharp.

I will not promise that the API won't break at the moment, however it has been working fairly well with the few samples I've tried.

Feedback is also welcome, if you find some rough edges that would be great to know about.

**Comment by tintoy** [comment]
Ok, I'll pull them from your appveyor feed for now (and mirror that into a myget feed if the rate-of-change warrants it) but will switch to NuGet once you've published them.

I managed to get your sample with server and vscode extension working (had to enable C#7.1 language features), so it shouldn't be too hard to transfer that over to a new project.

Thanks for building this, BTW - you're making my life a lot easier.

**Comment by tintoy** [comment]
Hmm, ok - that's a bit odd. I'm having trouble pulling in `JsonRpc`:

```
 /Users/tintoy/development/github/tintoy/msbuild-project-tools-vscode/src/LanguageServer/LanguageServer.csproj : warning NU1701: Package 'JsonRPC 1.0.1' was restored using '.NETPortable,Version=v0.0,Profile=win+net40+sl40+wp71, .NETFramework,Version=v4.6.1' instead the project target framework '.NETCoreApp,Version=v2.0'. This may cause compatibility problems.
/Users/tintoy/development/github/tintoy/msbuild-project-tools-vscode/src/LanguageServer/LanguageServer.csproj : warning NU1701: Package 'Microsoft.Bcl 1.0.19' was restored using '.NETPortable,Version=v0.0,Profile=win+net40+sl40+wp71, .NETFramework,Version=v4.6.1' instead the project target framework '.NETCoreApp,Version=v2.0'. This may cause compatibility problems.
/Users/tintoy/development/github/tintoy/msbuild-project-tools-vscode/src/LanguageServer/LanguageServer.csproj : warning NU1701: Package 'Microsoft.Bcl.Async 1.0.16' was restored using '.NETPortable,Version=v0.0,Profile=win+net40+sl40+wp71, .NETFramework,Version=v4.6.1' instead the project target framework '.NETCoreApp,Version=v2.0'. This may cause compatibility problems.
  Restore completed in 27.02 ms for /Users/tintoy/development/github/tintoy/msbuild-project-tools-vscode/src/LanguageServer/LanguageServer.csproj.
Done: 0.
```

Looking at the project file, it targets netstandard1.6, but the NuGet package in the feed says it references `Microsoft.Bcl.Async` (which is not listed in the project file, so I have no idea where it's getting that from). I might just build the packages locally for now. @Lercher have you ever seen this issue before? I'm targeting .NET Core 2.0, if it makes any difference.

**Comment by Lercher** [comment]
(1) I'm afraid to admit that I'm a bit confused with the targeting topic in general. So I can only provide this information: I used Visual Studio 2017 to build it locally, i.e. net46 and net standard 1.6, on a May '17 version. The build log, abbreviated:
````
1>JsonRpc -> C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\bin\Debug\net46\JsonRpc.dll
1>JsonRpc -> C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\bin\Debug\netstandard1.6\JsonRpc.dll
````

Just trieda quick `dotnet restore` / `dotnet build` with a result of 6W/0E with my old codebase. @tintoy Are you positive that you restored *this* solution? Could be my old code base, but I don't have a LanguageServer subdirectory at all. See my full log:

````
C:\Daten\GitHub\csharp-language-server-protocol>dotnet restore
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\sample\SampleServer\SampleServer.csproj...
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\JsonRpc.csproj...
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\JsonRpc.Tests.csproj...
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\src\Lsp\Lsp.csproj...
  Generating MSBuild file C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\obj\JsonRpc.csproj.nuget.g.props.
  Generating MSBuild file C:\Daten\GitHub\csharp-language-server-protocol\src\Lsp\obj\Lsp.csproj.nuget.g.props.
  Writing lock file to disk. Path: C:\Daten\GitHub\csharp-language-server-protocol\src\Lsp\obj\project.assets.json
  Writing lock file to disk. Path: C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\obj\project.assets.json
  Restore completed in 526,37 ms for C:\Daten\GitHub\csharp-language-server-protocol\src\JsonRpc\JsonRpc.csproj.
  Restore completed in 526,58 ms for C:\Daten\GitHub\csharp-language-server-protocol\src\Lsp\Lsp.csproj.
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\JsonRpc.Tests.csproj...
  Restoring packages for C:\Daten\GitHub\csharp-language-server-protocol\test\Lsp.Tests\Lsp.Tests.csproj...
  Generating MSBuild file C:\Daten\GitHub\csharp-language-server-protocol\sample\SampleServer\obj\SampleServer.csproj.nuget.g.props.
  Writing lock file to disk. Path: C:\Daten\GitHub\csharp-language-server-protocol\sample\SampleServer\obj\project.assets.json
  Restore completed in 812,32 ms for C:\Daten\GitHub\csharp-language-server-protocol\sample\SampleServer\SampleServer.csproj.
  Generating MSBuild file C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\obj\JsonRpc.Tests.csproj.nuget.g.props.
  Writing lock file to disk. Path: C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\obj\project.assets.json
  Restore completed in 886,65 ms for C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\JsonRpc.Tests.csproj.
  Generating MSBuild file C:\Daten\GitHub\csharp-language-server-protocol\test\Lsp.Tests\obj\Lsp.Tests.csproj.nuget.g.props.
  Writing lock file to disk. Path: C:\Daten\GitHub\csharp-language-server-protocol\test\Lsp.Tests\obj\project.assets.json
  Restore completed in 465,28 ms for C:\Daten\GitHub\csharp-language-server-protocol\test\Lsp.Tests\Lsp.Tests.csproj.
  Restore completed in 540,15 ms for C:\Daten\GitHub\csharp-language-server-protocol\test\JsonRpc.Tests\JsonRpc.Tests.csproj.

  NuGet Config files used:
      C:\Daten\GitHub\csharp-language-server-protocol\NuGet.Config
      C:\Users\lercher\AppData\Roaming\NuGet\NuGet.Config
      C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

  Feeds used:
      https://api.nuget.org/v3/index.json
      C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\
````


(2) I found that the lib works stable on my machine, including non ASCII sources. However, in May there was the plan to Opt In to do some language server work in parallel. I didn't use this option.

(3) I'm currently very busy working for my employer, so I had to postpone the work on LSP (that I currently consider done, because it's stable) and my Coco/R integration. So, sorry for my old sources.

**Comment by tintoy** [comment]
No worries - thanks for the detailed reply :)

I built the packages locally targeting netstandard1.6 only, used those, and
it works!

**Comment by tintoy** [comment]
@Lercher @david-driscoll it's  **ALIVE**:

![screenshot](https://user-images.githubusercontent.com/1646284/29291037-9a07cff8-8185-11e7-8747-685c662fc380.jpg)

**Comment by tintoy** [comment]
[Also:](https://github.com/tintoy/msbuild-project-tools-vscode/blob/f4e95a2d27d0368bc0fb8d433bb25f32e7d96613/src/LanguageServer.Engine/Handlers/ProjectDocumentHandler.cs#L126)

![msbuildlanguageserviceinaction](https://user-images.githubusercontent.com/1646284/29300155-2041d858-81b7-11e7-8ce3-d8bf1fb51e26.gif)

**Comment by Lercher** [comment]
Cool!

**Comment by tintoy** [comment]
Now I just need to work out how the MSBuild engine has changed in this new world (complains about being unable to find `Microsoft.NET.SDK`).

**Comment by tintoy** [comment]
I've got it all working, so I'm going to close this issue - I'll keep an eye on the repository though so when the LSP packages are ready for prime-time, I'll use those instead :)
